### ISSUE TITLE
Ranking  for function signatures.

### ISSUE THREAD

**Issue by pipermerriam** [ethereum-lists/4bytes#23]
Here's a starting point for how we *might* objectively rank function signatures.

- Let `C` be the set of all contract addresses which contain bytecode matching the pattern used to `JUMPDEST` based on the first 4 bytes of the message data.

With just `C` we can establish a basic ranking for signatures.  This ranking is however trivial to game.

- Let `T` be the set of all transactions who's first 4 bytes  match the signature.

With `len(T)` or `sum(t.gas_price * t.gas for t in T)` we should have a less easy to game metric.  I suspect that this will be suitable until we find someone directly attacking the rankings at which point we can iterate on this.

Question is, how do we easily get these metrics.  I think there is a BigQuery database for most of the chain data that I may be able to get access to, otherwise, maybe someone else knows of a relational database with all the chain data?

**Comment by pipermerriam** [comment]
cc @holiman @ligi

**Comment by holiman** [comment]
Why rank them? What's the objective? And why not just look at all method calso in the last N (100k) blocks and produce a toplist? That would also show the size of supplied data, which can be checked if it matches the abi parameters

**Comment by TrevorJTClarke** [comment]
Seems like a cool idea! Speaking internally (Amberdata). Thanks for reaching out on twitter!
I could see a couple useful cases however, esp if you identified top methods that cause issues/security flaws.

**Comment by ligi** [comment]
> Why rank them?

I think ranking them is a great way to know which one to show as a default. Perhaps even hide the lower ranked one when the difference in rank is really big (e.g. in the case of the ERC-20 forced collisions)

**Comment by holiman** [comment]
To weed out collisions, though... Then you need to validate calldata against abi params. And then the collision-game just becomes a bit harder since they need to use identical params. 

Or would you use natspec to resolve conflict?

**Comment by holiman** [comment]
So what I meant with 'why rank' is that it won't solve collisions, not easily at least.

**Comment by holiman** [comment]
The whole thing can be done pretty nicely with geth tracing, though. Except the natspec part, I guess.

**Comment by pipermerriam** [comment]
@holiman my thought is that 

1. ranking *can* at least partially solve artificial/malicious collisions.
2. all things being equal (selector and function params), allows for ordering based on a metric for which is most *likely* to be the desired one.

**Comment by holiman** [comment]
Re 2). How? How can you tell which sig is the one being called?

**Comment by pipermerriam** [comment]
@holiman I don't think we can *know*, but my thought is that we're just providing some contextual data and it's up to people who use this list to decide what to do with it.

**Theoretically** EthPM is the real fix for what this list is used for, at least in the case of wallets.  If people use packages as their gateway to accessing contracts, then the wallet can just use the ABI data from the package instead of having to use a possibly wrong reverse lookup from this list.

**Comment by holiman** [comment]
Oh wow. I remembered having implemented a 4byte-tracer long ago, and thought it may have gotten lost since then. But it's actually sitting right here: https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/internal/tracers/call_tracer.js 

So it's fairly trivial to just let it run on a few thousand blocks, and we can see how good coverage 4byte has right now, and what the big blind spots are.

**Comment by holiman** [comment]
Usage
```
> debug.traceTransaction( "0x214e597e35da083692f5386141e69f47e973b2c56e7a8073b1ea08fd7571e9de", {tracer: "4byteTracer"})
   {
     0x27dc297e-128: 1,
     0x38cc4831-0: 2,
     0x524f3889-96: 1,
     0xadf59f99-288: 1,
     0xc281d19e-0: 1
   }
```
It's also possible to trace a range of blocks, and continually get pushes when new blocks arrive, if you use ipc+subscription.

**Comment by holiman** [comment]
@pipermerriam , you wanted to weigh in the gas and gasprice. Do you really that is needed?
Also, I still don't see how this will solve anything regarding duplicates, but I think it's good stats to have to know measure how good coverage the directory has.

**Comment by ligi** [comment]
@holiman :+1: 

I do not think the gas and gas-price is needed.
Can I have the full list somewhere? With the calldata-size I can already resolve some conflicts - e.g. with:
`available_assert_time(uint16,uint64);decimals() `

so `1395 0x313ce567-0` shows me that decimals() is the most likely candidate

but was just thinking if there is a collision that is really not forced but done by chance - then this is really problematic. This method will never really have the chance to change it's position if we hide it in the dark. 
Perhaps @yann300 can show a warning in remix if a function signature resolves in a known 4byte? 

In the end we really need natspec - and then shame contracts that did not upload a natspec to swarm by making a big red warning sign on all transactions that interact with such contracts.

**Comment by kumavis** [comment]
for comparing 4byte collisions, preferring the one with the simplest signature (shortest name, fewest arguments) is a decent heuristic
